### DISCUSSION TITLE
ES Modules support with clientlibs

### DISCUSSION THREAD

**Discussion by jamesg1** [aem-vite/aem-vite#13] [Q&A] [answered]
Hey,

Firstly great library and I've enjoyed using it with some recent aem 6.5 upgrade work. Vite is a significant improvement to our development experience.

I wanted to ask how you've managed to load clientlibs with `type="module"` within aem 6.5 and if you've had to do anything fancy.

I am using on prem AEM 6.3/6.5 and using versioned clientlibs to add a unique hash onto the clientlibs - running into this https://github.com/Adobe-Consulting-Services/acs-aem-commons/issues/2560 due to transformers not liking scripts when its not`text/javascript`.

My next idea is to add a custom transformer but wanted to check in and see if there was another way.

Thanks!

**Comment by jamesg1** [comment]
Ended up adding support for versioned clientlibs here https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/2691

**Comment by cshawaus** [comment, marked as answer]
Hi @jamesg1,

Apologies for the very belated reply. As part of the installation process, there is a step to switch your current ClientLib HTL use attribute over to the AEM Vite ClientLib handler.

https://www.aemvite.dev/guide/backend/installation/#configure-clientlibs

Of course, this is not always optimal but you can copy the files from `/apps/aem-vite/granite/sightly/templates` into your project folder and extend https://github.com/aem-vite/aem-vite/blob/main/core/src/main/java/dev/aemvite/aem/utilities/ClientLibUseObject.java.

This should play nicely with Versioned ClientLibs as it executes before the rewriter.

**Reply by jamesg1** [reply]
Thanks will give that a crack and let you know.

**Reply by sachinlala** [reply]
Hi @cshawaus,

AEM Vite is a great initiative and I have been working on migrating the AEMaaCS archtype we have.

I need your help for the following issue - 

> 05.08.2024 08:20:39.076 *ERROR* [[0:0:0:0:0:0:0:1] [1722826237316] GET /content/bhn/language-masters/en.html HTTP/1.1] com.day.cq.wcm.core.impl.WCMDeveloperModeFilter Error during include of SlingRequestPathInfo: path='/content/bhn/language-masters/en/jcr:content', selectorString='null', extension='html', suffix='null'
org.apache.sling.scripting.sightly.SightlyException: Identifier /apps/aem-vite/granite/sightly/templates/clientlib.html cannot be correctly instantiated by the Use API

> Caused by: org.apache.sling.scripting.sightly.SightlyException: Cannot find resource /apps/aem-vite/granite/sightly/templates/clientlib.html for base path /apps/bhn/components/page with resource super type core/cif/components/structure/page/v3/page.

I have tried to follow the steps diligently, as mentioned in https://www.aemvite.dev/guide/backend/clientlibs/ and also (for now) manually copied-over the `clientlib.html` and `graniteClientLib.html` files from your repo to the same folder-structure inside `ui.apps` module. Further on, have also included the new path in `vault/filter.xml` -> `<filter root="/apps/aem-vite"/>`.

But still getting the error above.

Please can you suggest what I could be missing here ? Happy to provide more information as needed. 

Also, I'd like to avoid copying over the 2 files and rather reference them from the aem-vite pom artifact.
